pkg/ingester/ingester.go

```diff
@@ -739,6 +832,47 @@ func (i *Ingester) MetricsForLabelMatchers(ctx context.Context, req *client.Metr
 	return result, nil
 }
 
+// MetricMetadata returns all the metric metadata of a user.
+func (i *Ingester) MetricMetadata(ctx context.Context, req *client.MetricMetadataRequest) (*client.MetricMetadataResponse, error) {
+	i.userStatesMtx.RLock()
+	if err := i.checkRunningOrStopping(); err != nil {
+		i.userStatesMtx.RUnlock()
+		return nil, err
+	}
+	i.userStatesMtx.RUnlock()
+
+	userID, err := user.ExtractOrgID(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("no user id")
+	}
+
+	i.usersMetadataMtx.RLock()
+	defer i.usersMetadataMtx.RUnlock()
+	userMetadata, ok := i.usersMetadata[userID]
+	if !ok {
+		return &client.MetricMetadataResponse{}, nil
+	}
+
+	if userMetadata.length() > i.limiter.MaxMetricsWithMetadataPerQuery(userID) {
+		return nil, httpgrpc.Errorf(http.StatusRequestEntityTooLarge, "exceeded maximum number of metadata metrics in a query")
+	}
```

Comment:
Per private discussion between you and Marco, I expect that limits on read paths will be removed (?)

Reply:
Limits have been removed.